### PR TITLE
New exercise: Tower of Hanoi.

### PR DESCRIPTION
**Pull request by udhos**
See https://github.com/exercism/go/issues/1270

### CODE REVIEW

**Comment by bitfield** [comment]
Would one way to simplify this exercise slightly be to assume that the rods are always called "A", "B", and "C"? Then the solution program no longer needs those inputs, and the only input would be the number of disks.

**Comment by bitfield** [comment]
Yes, I see now, this is a pretty standard thing with Towers of Hanoi solvers. How about calling them `from`, `to`, and `via`, then?

**Comment by bitfield** [comment]
It doesn't seem very Go-like to use strings for the rod names. How about defining constants:

```go
const (
	From = iota
	To
	Via
)
```

**Comment by udhos** [comment]
Pushed a change replacing rods names with numeric IDs.

**Comment by bitfield** [comment]
I'm not sure the numeric IDs are really an improvement over A, B, and C. Given that the key insight in solving this puzzle is swapping around the From, To, and Via rods in recursive calls to Solve, shouldn't we name them `From`, `To`, and `Via` as suggested? If you can come up with better names, by all means do!

**Comment by udhos** [comment]
Well, the func parameters were named 'from', 'to', 'via' already, but their type was string. I thought the point of defining numeric constants was to change their type from string id to numeric id.

So I think I am misunderstanding something.

Can you show the interface you have in mind for the function?

**Comment by bitfield** [comment]
Yes, absolutely: the signature is fine, but I'm suggesting we use the named constants instead of 10, 20, 30, etc. Actually, now that I understand the problem a bit more, I think the names A, B, and C make sense:

```
Your implementation will accept the amount of disks on rod A, then issue a list of moves for solving the puzzle.

Example:

Input:
- number of disks: 2

Output: list of three moves: {A,C}, {A,B}, {C,B} 
```

**Comment by udhos** [comment]
Oh, now I see what you mean: names used in the exercise description.
I am trying to figure how to use the constants in the problem description while not suggesting to the student that those constants would actually help in the implementation, since the recursive solution itself would not benefit from them.
Please see a sketch below.
What do you think?

```
Your implementation will accept the amount of disks on rod A, then issue a list of moves for solving the puzzle.

// Consider the constants below naming the rods, for the sake of the problem description.
// Your implementation is free to define similar constants, but this is not a requirement.
const (
	A = iota
	B
	C
)

Example:

Input:
- number of disks: 2

Output: list of three moves: {A,C}, {A,B}, {C,B}
```

**Comment by bitfield** [comment]
Yes, for the sake of generating readable output, it's maybe better to define, for example:

```go
const A = 'A'
...

```

What do you think?

**Comment by udhos** [comment]
I added pretty printing to tests.
Now we can keep rods names like A,B,C in the generic description, and tests would report similar names.

go test passing:
```
    hanoi_test.go:80: disks=7 move=[A B] before: {[1] [7 6 5 4 3 2] []}
    hanoi_test.go:140: disks=7 move=[A B] after: {[] [7 6 5 4 3 2 1] []}
    hanoi_test.go:31: PASS totalDisks=7
```

go test failing:
```
    hanoi_test.go:80: disks=7 move=[A B] before: {[7 6 5 4 3 2 1] [] []}
    hanoi_test.go:140: disks=7 move=[A B] after: {[7 6 5 4 3 2] [1] []}
    hanoi_test.go:28: FAIL totalDisks=7 rod TO=B has 1 disks, should have 7: [1]
```

In Go we state:

```
// Example
// Solve(1, 'A', 'B', 'C') could return a list with a single move:
//  [][2]int{{'A', 'B'}} // move disk from rod 'A' to rod 'B'
```

**Comment by udhos** [comment]
I am unsure about how much information we should put in the stub file 'hanoi.go'.

Can you guys please have a look at how it looks?

**Comment by bitfield** [comment]
How come you closed this, @udhos?

**Comment by udhos** [comment]
Sorry, I guess I closed it by accident.

**Comment by leenipper** [comment]
> how much information we should put in the stub file 'hanoi.go'

It looks ok I think.  @bitfield ?

File: exercises/tower-of-hanoi/hanoi.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package towerofhanoi
+
+// This is an stub file. Remove this comment and
```
Comment (bitfield) [inline exercises/tower-of-hanoi/hanoi.go]:
```suggestion
// This is a stub file...
```

File: exercises/tower-of-hanoi/hanoi.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// implement the function Solve below.
+
+// Solve returns list of moves for solving the Tower of Hanoi puzzle.
+//
+// 'disks' is the amount of disks.
+// 'from' is the id for the source rod.
+// 'to' is the id for the destination rod.
+// 'via' is the id for the helper rod.
+//
+// A move is a [2]int value specifying source and destination rods.
```
Comment (bitfield) [inline exercises/tower-of-hanoi/hanoi.go]: This strikes me as a little weird. I've literally never used an array in Go. Would a struct `{from, to int}` be a bit more idiomatic?

File: exercises/tower-of-hanoi/hanoi.go (hunk 1 above)
Comment (udhos) [reply exercises/tower-of-hanoi/hanoi.go]: Added as https://github.com/exercism/go/pull/1272/commits/bc507303b28de3db175928d057303032d4b8c3f2

Unsure whether you suggested the full change or just the grammar issue.

File: exercises/tower-of-hanoi/hanoi.go (hunk 2 above)
Comment (udhos) [reply exercises/tower-of-hanoi/hanoi.go]: Please see https://github.com/exercism/go/pull/1272/commits/1741aab3426a00e8173927631eeed43300ab8b13

**Comment by CoderYihaoWang** [comment]
This looks like a very interesting exercise and as a learner I would love to see it on the website. But it seems that it has been ignored since the last modification . Is this abandoned? @leenipper @bitfield @udhos

**Comment by bitfield** [comment]
> Is this abandoned?

So it seems!

**Comment by jmrunkle** [comment]
I am closing all PRs that are over a year old. Please reopen if you are still working on this.

**Pull request closed without merging**